File: samples/test_dma_map.c
/*
 * Userspace mediated device sample application
 *
 * Copyright (c) 2019, Nutanix Inc. All rights reserved.
 *     Author: Thanos Makatos <thanos@nutanix.com>
 *             Swapnil Ingle <swapnil.ingle@nutanix.com>
 *             Felipe Franciosi <felipe@nutanix.com>
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *      * Redistributions of source code must retain the above copyright
 *        notice, this list of conditions and the following disclaimer.
 *      * Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 *      * Neither the name of Nutanix nor the names of its contributors may be
 *        used to endorse or promote products derived from this software without
 *        specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
 *  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 *  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 *  OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 *  DAMAGE.
 *
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <stdbool.h>
#include <string.h>
#include <linux/vfio.h>
#include <limits.h>
#include <assert.h>
#include <sys/ioctl.h>
#include <inttypes.h>
#include <err.h>
#include "../lib/muser_priv.h"

#define ARRAY_SIZE(array) (sizeof(array) / sizeof(*array))

#define VFIO_PATH           "/dev/vfio/"
#define VFIO_CTR_PATH       VFIO_PATH "vfio"
#define SYSFS_PCI_DEV_PATH  "/sys/bus/pci/devices/"
#define SYSFS_IOMMU_GROUP   "/iommu_group"

static int
pci_group_id(const char *bdf)
{
	char *dev_path;
	char group_path[PATH_MAX];
	int group_id;

	assert(bdf);

	asprintf(&dev_path, SYSFS_PCI_DEV_PATH "%s" SYSFS_IOMMU_GROUP, bdf);
	memset(group_path, 0, sizeof(group_path));
	readlink(dev_path, group_path, sizeof(group_path));
	free(dev_path);
	sscanf(basename(group_path), "%d", &group_id);
	return group_id;
}

static void*
test_map_dma(const int fd, void *vaddr, unsigned long size, unsigned long iova)
{
	int err;
	struct vfio_iommu_type1_dma_map dma_map = {
		.argsz = sizeof(dma_map),
		.flags = VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE,
		.vaddr = (unsigned long long)vaddr,
		.iova = iova,
		.size = size,
	};
	fprintf(stderr, "attempting to MAP_DMA IOVA=%llx\n", dma_map.iova);

	err = ioctl(fd, VFIO_IOMMU_MAP_DMA, &dma_map);
	if (err) {
		fprintf(stderr, "failed to MAP_DMA: %d (%s)\n", err,
		        strerror(errno));
		return NULL;
	}

	return (void*)dma_map.vaddr;
}

static void
test_unmap_dma(const int fd, unsigned long size, unsigned long long iova)
{
	int err;
	struct vfio_iommu_type1_dma_unmap dma_unmap = {
		.argsz = sizeof dma_unmap,
		.size = size,
		.iova = iova,
		.flags = 0
	};

	err = ioctl(fd, VFIO_IOMMU_UNMAP_DMA, &dma_unmap);
	if (err) {
		perror("UNMAP_DMA\n");
		return;
	}
	printf("unmapped IOVA=%llx\n", dma_unmap.iova);
}

static int
get_container_fd(const char *path)
{
	int err, vfio_ctr_fd, vfio_grp_fd, vfio_dev_fd;
	char *grp_path;

	vfio_ctr_fd = open(VFIO_CTR_PATH, O_RDWR);
	assert(vfio_ctr_fd >= 0);

	// Open the VFIO entry for this device's IOMMU GROUP.
	err = asprintf(&grp_path, VFIO_PATH "%d", pci_group_id(path));
	assert(err > 0);
	vfio_grp_fd = open(grp_path, O_RDWR);
	assert(vfio_grp_fd >= 0);
	free(grp_path);

	// Add the group to the container.
	err = ioctl(vfio_grp_fd, VFIO_GROUP_SET_CONTAINER, &vfio_ctr_fd);
	assert(!err);

	// Enable IOMMU type 1 on container.
	err = ioctl(vfio_ctr_fd, VFIO_SET_IOMMU, VFIO_TYPE1v2_IOMMU);
	assert(!err);

	// Get a device fd from VFIO.
	vfio_dev_fd = ioctl(vfio_grp_fd, VFIO_GROUP_GET_DEVICE_FD, path);
	assert(vfio_dev_fd >= 0);

	return vfio_ctr_fd;
}

int main(int argc, char * argv[])
{
	int vfio_ctr_fd;
	void *dma_map_addr;
	struct iovec dma_regions[] = {
		{.iov_base = (void*)0x0, .iov_len = 1 << 21},
		{.iov_base = (void*)(1 << 21), .iov_len = 1 << 21},
	};
	size_t i;
	bool huge = true;
	int fd;
	int flags = MAP_SHARED;
	void *vaddr;
	size_t size = 1 << 23; /* FIXME */
	int ret;

	if (argc != 2) {
		fprintf(stderr, "usage: %s <MUSER device UUID>\n", argv[0]);
		return EXIT_FAILURE;
	}

	vfio_ctr_fd = get_container_fd(argv[1]);

	if (huge) {
		char template[] = "/dev/hugepages/XXXXXX";
		fd = mkstemp(template);
		assert(fd != -1);
	}

	ret = lseek(fd, size, SEEK_END);
	if (ret == -1) {
		err(EXIT_FAILURE, "failed to seek at %lu", size);
	}

	/*
	 * Allocate some space and setup a DMA mapping.
	 * It *must* be MAP_SHARED.
	 */
	vaddr = mmap(NULL, size, PROT_READ | PROT_WRITE, flags, fd, 0);
	if (vaddr == MAP_FAILED) {
		err(EXIT_FAILURE, "failed to mmap");
	}

	for (i = 0; i < ARRAY_SIZE(dma_regions); i++) {
		if ((unsigned long long)dma_regions[i].iov_base + dma_regions[i].iov_len > size
		    || (huge && (dma_regions[i].iov_len < (1 << 21)))) {
			err(EXIT_FAILURE, "bad IOVA %#lx-%#lx\n",
			    (unsigned long)dma_regions[i].iov_base,
			    (unsigned long)dma_regions[i].iov_base + dma_regions[i].iov_len);
		}
		dma_map_addr = test_map_dma(vfio_ctr_fd, vaddr,
		                            dma_regions[i].iov_len,
		                            (unsigned long)dma_regions[i].iov_base);
		if (!dma_map_addr)
			exit(EXIT_FAILURE);
	}

	printf("press enter to continue\n");
	getchar();

	for (i = 0; i < ARRAY_SIZE(dma_regions); i++) {
		test_unmap_dma(vfio_ctr_fd,
		               (unsigned long long)dma_regions[i].iov_len,
		               (unsigned long long)dma_regions[i].iov_base);
	}
	return 0;
}

File: test/py/test_dma_map.py
#
# Copyright (c) 2021 Nutanix Inc. All rights reserved.
#
# Authors: John Levon <john.levon@nutanix.com>
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
#      * Redistributions of source code must retain the above copyright
#        notice, this list of conditions and the following disclaimer.
#      * Redistributions in binary form must reproduce the above copyright
#        notice, this list of conditions and the following disclaimer in the
#        documentation and/or other materials provided with the distribution.
#      * Neither the name of Nutanix nor the names of its contributors may be
#        used to endorse or promote products derived from this software without
#        specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
#  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
#  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
#  SERVICESLOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
#  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
#  OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
#  DAMAGE.
#

from unittest import mock
from unittest.mock import patch, Mock, create_autospec

from libvfio_user import *
import errno

#
# NB: this is currently very incomplete
#

ctx = None
quiesce_cb_err = 0
mock_reset_cb = Mock(return_value=0)


@vfu_reset_cb_t
def reset_cb(ctx, reset_type):
    global mock_reset_cb
    return mock_reset_cb(ctx, reset_type)


@vfu_device_quiesce_cb_t
def quiesce_cb(ctx):
    global quiesce_cb_err
    if quiesce_cb_err != 0:
        c.set_errno(quiesce_cb_err)
        return -1
    return 0


def setup_function(function):
    global mock_quiesce_cb, ctx, sock
    mock_reset_cb.reset_mock()
    ctx = prepare_ctx_for_dma(quiesce=quiesce_cb, reset=reset_cb)
    assert ctx is not None
    sock = connect_client(ctx)


def teardown_function(function):
    global mock_quiesce_cb, ctx, sock, quiesce_cb_err
    disconnect_client(ctx, sock)
    quiesce_cb_err = 0
    mock_reset_cb.return_value = 0
    ret = vfu_destroy_ctx(ctx)
    assert ret == 0, "failed to destroy context, ret=%s, errno=%s" % (ret, c.get_errno())


def test_dma_region_too_big():
    global ctx, sock

    payload = vfio_user_dma_map(argsz=len(vfio_user_dma_map()),
        flags=(VFIO_USER_F_DMA_REGION_READ |
               VFIO_USER_F_DMA_REGION_WRITE),
        offset=0, addr=0x10000, size=MAX_DMA_SIZE + 4096)

    msg(ctx, sock, VFIO_USER_DMA_MAP, payload, expect=errno.ENOSPC)




def test_dma_region_too_many():
    global ctx, sock

    for i in range(1, MAX_DMA_REGIONS + 2):
        payload = vfio_user_dma_map(argsz=len(vfio_user_dma_map()),
            flags=(VFIO_USER_F_DMA_REGION_READ |
                   VFIO_USER_F_DMA_REGION_WRITE),
            offset=0, addr=0x1000 * i, size=4096)

        if i == MAX_DMA_REGIONS + 1:
            expect = errno.EINVAL
        else:
            expect = 0

        msg(ctx, sock, VFIO_USER_DMA_MAP, payload, expect=expect)


def test_dma_map_busy():
    global ctx, sock, quiesce_cb_err
    quiesce_cb_err = errno.EBUSY

    payload = vfio_user_dma_map(argsz=len(vfio_user_dma_map()),
        flags=(VFIO_USER_F_DMA_REGION_READ |
               VFIO_USER_F_DMA_REGION_WRITE),
        offset=0, addr=0x10000, size=0x1000)

    msg(ctx, sock, VFIO_USER_DMA_MAP, payload, rsp=False)

    ret = vfu_run_ctx(ctx)
    assert ret == -1
    assert c.get_errno() == errno.EBUSY

    vfu_device_quiesced(ctx, 0)

    dma_register_cb_err = 0

    get_reply(sock)

    ret = vfu_run_ctx(ctx)
    assert ret == 0


# FIXME need the same test for (1) DMA unmap, (2) device reset, and
# (3) migration, where quiesce returns EBUSY but replying fails.
def test_dma_map_busy_reply_fail():
    """Tests mapping a DMA region where the quiesce callback returns EBUSY and replying fails."""

    global ctx, sock, quiesce_cb_err

    # Send a DMA map command.
    payload = vfio_user_dma_map(argsz=len(vfio_user_dma_map()),
        flags=(VFIO_USER_F_DMA_REGION_READ |
               VFIO_USER_F_DMA_REGION_WRITE),
        offset=0, addr=0x10000, size=0x1000)

    # device will be busy quiescing
    quiesce_cb_err = errno.EBUSY

    msg(ctx, sock, VFIO_USER_DMA_MAP, payload, rsp=False)

    ret = vfu_run_ctx(ctx)
    assert ret == -1
    assert c.get_errno() == errno.EBUSY

    # TODO check that quiesce has been called

    # pretend there's a connection failure while the device is still quiescing
    sock.close()

    # device reset callback should not have been called so far
    assert mock_reset_cb.call_count == 0

    # device quiesces
    vfu_device_quiesced(ctx, 0)

    # device reset callback should be called
    mock_reset_cb.assert_has_calls([mock.call(ctx, True)])

    ret = vfu_run_ctx(ctx)
    assert ret == -1
    assert c.get_errno() == errno.ENOTCONN


# ex: set tabstop=4 shiftwidth=4 softtabstop=4 expandtab: #
